### PR TITLE
Starter theme/fix/template override

### PR DESCRIPTION
**Pull request by julienbourdeau**
## Custom templates

We distinguished layout and template:
- **Layout**: The way the page is arranged (full width, 2 columns, content only, etc)
- **Template**: The actual smarty file loaded: `cms/page.tpl`, `catalog/product.tpl`

This seems pretty obvious but the code wasn't exactly working this way.
PrestaShop used to call it _theme overrides_ now there is no `override/` folder anymore, only specific template.

FrontControllers usually assign template this way:

``` php
$this->setTemplate('catalog/product.tpl');
```

and we load the file found at this exact location in the theme.

Now `setTemplate()` we look for more specific templates before. As of today (for the `catalog/product.tpl` example):
1. catalog/product-{id_product}.tpl
2. catalog/product.tpl

**Do you have any idea for extra rules ?**

**Note:** _product-{id_product}.tpl_ can extend _product.tpl_ and simply override one smart block.
## Module Smarty Resource for templates

This is some sort of namespace for module template. Instead of loading
a file from the given path, Smarty will look for a corresponding template
in multiple locations (following priority order). Thanks to [Smarty Custom Resources](http://www.smarty.net/docs/en/resources.custom.tpl).

PrestaShop already allow theme to override module templates but if they
were using `include` statement, you had to override theme all.

Now module can allow theme to override only one template by using the following
syntax.

``` smarty
{include file="module:module_directory_name/path/to/template/file.tpl"}
```
### Example with blockcart module

If we want to allow theme to override only the product line template
but not the main one (and vice versa), you simply need to replace:

``` smarty
<li>{include './blockcart-product-line.tpl' product=$product}</li>
```

by

``` smarty
<li>{include file="module:blockcart/blockcart-product-line.tpl" product=$product}</li>
```

### CODE REVIEW

**Comment by kpodemski** [comment]
Hmmm... so, if we have blockcart.tpl provided by  native PrestaShop module in: modules/blockcart/views/templates/hook/blockcart.tpl, and this .tpl file is used in `displayTop` how can we now override it? We are still using `template/override` but inside this override we're using special version of `{include}`, right?

**Comment by julienbourdeau** [comment]
Hi,

About templates base on id_shop, I thought about is, but quickly I realized it would introduce too much complexity and I don't think it will be used so much.

For your override, if the template is set by php (using `$this->display()`) it will first look in your theme: `_PS_THEME_DIR_.'/modules/blockcart/views/templates/hook/blockcart.tpl`. It's the same behavior as PS1.6. Did I answer your question ?

**Comment by kpodemski** [comment]
Hi @julienbourdeau 

Yes, `id_shop` is too much, and yes - your answer is ok for me. Will you move this further? I mean, maybe if we have installed `mytheme`, we can look for `mytheme-child` and if `product.tpl` exsits in child theme it has highest priority? Of course we'll need additional solution for `{include}` similar to this one for modules. What you think?

**Comment by julienbourdeau** [comment]
We will definitely deliver a parent/child system for theme. Yeah, I guess this include system will evolve with it.

**Comment by kpodemski** [comment]
:+1:

**Comment by djfm** [comment]
`We will definitely deliver a parent/child system for theme.`, yeah, but not right now :)

**Comment by kpodemski** [comment]
@djfm but what is the most hard thing to do here? Is overriding directory path + `{include}` based on current work is not good enough?

File: config/defines_uri.inc.php [hunk 1]
```diff
@@ -32,7 +32,7 @@
 define('_THEME_IMG_DIR_', _THEME_DIR_.'assets/img/');
 define('_THEME_CSS_DIR_', _THEME_DIR_.'assets/css/');
 define('_THEME_JS_DIR_', _THEME_DIR_.'assets/js/');
-define('_PS_THEME_OVERRIDE_DIR_', _PS_THEME_DIR_.'override/');
+define('_PS_THEME_OVERRIDE_DIR_', _PS_THEME_DIR_.'templates/override/');
```
Comment (djfm) [inline config/defines_uri.inc.php]: I think we can remove this since we're not using it at all and we don't need backwards compatibility with themes.

**Pull request merged**